File: python/eventgraph/authority.py
"""Authority module — flat authority model with policy matching and trust downgrade.

Ports the Go authority package (DefaultAuthorityChain). Provides weighted
authority evaluation, not binary permission.
"""

from __future__ import annotations

import threading
from dataclasses import dataclass, field
from typing import Protocol, runtime_checkable

from .decision import AuthorityLevel
from .types import ActorID, DomainScope, EventID, Option, Score


# ── Protected Actions ────────────────────────────────────────────────────

PROTECTED_ACTION_PRODUCTION_DEPLOY = "production.deploy"
PROTECTED_ACTION_REPO_CREATE = "repo.create"
PROTECTED_ACTION_REPO_DELETE = "repo.delete"
PROTECTED_ACTION_REPO_PUSH_DEFAULT_BRANCH = "repo.push.default_branch"
PROTECTED_ACTION_REPO_MERGE_MAIN = "repo.merge.main"
PROTECTED_ACTION_REPO_MUTATE_CROSS_REPO = "repo.mutate.cross_repo"
PROTECTED_ACTION_SELF_MODIFICATION_ACTIVATE = "self_modification.activate"
PROTECTED_ACTION_SECRET_ACCESS = "secret.access"
PROTECTED_ACTION_POLICY_CHANGE = "policy.change"

PROTECTED_ACTIONS: tuple[str, ...] = (
    PROTECTED_ACTION_PRODUCTION_DEPLOY,
    PROTECTED_ACTION_REPO_CREATE,
    PROTECTED_ACTION_REPO_DELETE,
    PROTECTED_ACTION_REPO_PUSH_DEFAULT_BRANCH,
    PROTECTED_ACTION_REPO_MERGE_MAIN,
    PROTECTED_ACTION_REPO_MUTATE_CROSS_REPO,
    PROTECTED_ACTION_SELF_MODIFICATION_ACTIVATE,
    PROTECTED_ACTION_SECRET_ACCESS,
    PROTECTED_ACTION_POLICY_CHANGE,
)


def is_protected_action(action: str) -> bool:
    """Return True if *action* is a shared Dark Factory protected action."""
    return action in PROTECTED_ACTIONS


# ── Value Types ──────────────────────────────────────────────────────────


@dataclass(frozen=True, slots=True)
class AuthorityRequestContent:
    """Typed content for the authority.requested kernel event."""

    action: str
    actor: ActorID
    level: AuthorityLevel
    justification: str
    causes: tuple[EventID, ...]

    def __init__(
        self,
        action: str,
        actor: ActorID,
        level: AuthorityLevel,
        justification: str,
        causes: list[EventID],
    ) -> None:
        object.__setattr__(self, "action", action)
        object.__setattr__(self, "actor", actor)
        object.__setattr__(self, "level", level)
        object.__setattr__(self, "justification", justification)
        object.__setattr__(self, "causes", tuple(causes))

    def to_event_content(self) -> dict[str, object]:
        """Return canonical content keys for an authority.requested event."""
        return {
            "Action": self.action,
            "Actor": self.actor.value,
            "Level": self.level.value,
            "Justification": self.justification,
            "Causes": [cause.value for cause in self.causes],
        }


@dataclass(frozen=True, slots=True)
class AuthorityLink:
    """A link in an authority chain."""

    _actor: ActorID
    _level: AuthorityLevel
    _weight: Score

    def __init__(self, actor: ActorID, level: AuthorityLevel, weight: Score) -> None:
        object.__setattr__(self, "_actor", actor)
        object.__setattr__(self, "_level", level)
        object.__setattr__(self, "_weight", weight)

    @property
    def actor(self) -> ActorID:
        return self._actor

    @property
    def level(self) -> AuthorityLevel:
        return self._level

    @property
    def weight(self) -> Score:
        return self._weight


@dataclass(frozen=True, slots=True)
class AuthorityResult:
    """Result of evaluating authority for an action."""

    _level: AuthorityLevel
    _weight: Score
    _chain: tuple[AuthorityLink, ...]

    def __init__(
        self, level: AuthorityLevel, weight: Score, chain: list[AuthorityLink]
    ) -> None:
        object.__setattr__(self, "_level", level)
        object.__setattr__(self, "_weight", weight)
        object.__setattr__(self, "_chain", tuple(chain))

    @property
    def level(self) -> AuthorityLevel:
        return self._level

    @property
    def weight(self) -> Score:
        return self._weight

    @property
    def chain(self) -> list[AuthorityLink]:
        return list(self._chain)


@dataclass(frozen=True, slots=True)
class AuthorityPolicy:
    """Defines the authority requirements for an action pattern."""

    action: str
    level: AuthorityLevel
    min_trust: Option[Score] = field(default_factory=Option.none)
    scope: Option[DomainScope] = field(default_factory=Option.none)


# ── AuthorityChain Protocol ─────────────────────────────────────────────


@runtime_checkable
class AuthorityChain(Protocol):
    """Evaluates authority. Returns weighted authority, not binary permission."""

    def evaluate(self, actor: ActorID, action: str) -> AuthorityResult: ...

    def chain(self, actor: ActorID, action: str) -> list[AuthorityLink]: ...

    def grant(
        self,
        from_actor: ActorID,
        to_actor: ActorID,
        scope: DomainScope,
        weight: Score,
    ) -> None: ...

    def revoke(
        self, from_actor: ActorID, to_actor: ActorID, scope: DomainScope
    ) -> None: ...


# ── Helper ───────────────────────────────────────────────────────────────


def matches_action(pattern: str, action: str) -> bool:
    """Match an action against a pattern.

    - ``"*"`` matches everything.
    - ``"prefix*"`` matches any action starting with ``prefix``.
    - Otherwise exact match.
    """
    if pattern == "*":
        return True
    if len(pattern) > 0 and pattern[-1] == "*":
        prefix = pattern[:-1]
        return len(action) >= len(prefix) and action[: len(prefix)] == prefix
    return pattern == action


# ── DefaultAuthorityChain ────────────────────────────────────────────────


class DefaultAuthorityChain:
    """Flat authority model — no delegation chain.

    All actions default to Notification unless a policy says otherwise.
    If a Required policy has a ``min_trust`` threshold and the actor's trust
    meets or exceeds it, the level is downgraded to Recommended.

    Implements the :class:`AuthorityChain` protocol.
    """

    def __init__(self, trust_model: object) -> None:
        """Create a flat authority chain.

        Parameters
        ----------
        trust_model:
            An object implementing the ``TrustModel`` protocol (must have a
            ``score(actor) -> TrustMetrics`` method).
        """
        self._lock = threading.Lock()
        self._policies: list[AuthorityPolicy] = []
        self._trust_model = trust_model

    # ── Policy management ────────────────────────────────────────────────

    def add_policy(self, policy: AuthorityPolicy) -> None:
        """Register an authority policy. Policies are checked in order; first match wins."""
        with self._lock:
            self._policies.append(policy)

    # ── Private helpers ──────────────────────────────────────────────────

    def _find_policy(self, action: str) -> AuthorityPolicy:
        """Return the first matching policy, or the default Notification policy."""
        for p in self._policies:
            if matches_action(p.action, action):
                return p
        return AuthorityPolicy(action="*", level=AuthorityLevel.NOTIFICATION)

    # ── AuthorityChain protocol ──────────────────────────────────────────

    def evaluate(self, actor: ActorID, action: str) -> AuthorityResult:
        """Evaluate authority for *actor* performing *action*.

        If the matching policy is Required and specifies a ``min_trust``, and
        the actor's trust score meets or exceeds that threshold, the level is
        downgraded to Recommended.
        """
        with self._lock:
            policy = self._find_policy(action)

        level = policy.level

        # Trust-based downgrade: Required → Recommended
        if level == AuthorityLevel.REQUIRED and policy.min_trust.is_some():
            metrics = self._trust_model.score(actor)
            if metrics.overall.value >= policy.min_trust.unwrap().value:
                level = AuthorityLevel.RECOMMENDED

        link = AuthorityLink(actor=actor, level=level, weight=Score(1.0))

        return AuthorityResult(
            level=level,
            weight=Score(1.0),
            chain=[link],
        )

    def chain(self, actor: ActorID, action: str) -> list[AuthorityLink]:
        """Return the authority chain for *actor* performing *action*.

        In the flat model this is always a single-link chain.
        """
        with self._lock:
            policy = self._find_policy(action)

        return [AuthorityLink(actor=actor, level=policy.level, weight=Score(1.0))]

    def grant(
        self,
        from_actor: ActorID,
        to_actor: ActorID,
        scope: DomainScope,
        weight: Score,
    ) -> None:
        """Grant authority — no-op in the flat model.

        A full implementation would emit an ``edge.created`` event.
        """

    def revoke(
        self, from_actor: ActorID, to_actor: ActorID, scope: DomainScope
    ) -> None:
        """Revoke authority — no-op in the flat model.

        A full implementation would emit an ``edge.superseded`` event.
        """

File: dotnet/src/EventGraph/Authority.cs
namespace EventGraph;

// ── ProtectedAction ───────────────────────────────────────────────────

/// <summary>Dark Factory authority-gated protected action names from DF-SOP-0001.</summary>
public static class ProtectedAction
{
    public const string ProductionDeploy = "production.deploy";
    public const string RepoCreate = "repo.create";
    public const string RepoDelete = "repo.delete";
    public const string RepoPushDefaultBranch = "repo.push.default_branch";
    public const string RepoMergeMain = "repo.merge.main";
    public const string RepoMutateCrossRepo = "repo.mutate.cross_repo";
    public const string SelfModificationActivate = "self_modification.activate";
    public const string SecretAccess = "secret.access";
    public const string PolicyChange = "policy.change";

    public static IReadOnlyList<string> All { get; } = new[]
    {
        ProductionDeploy,
        RepoCreate,
        RepoDelete,
        RepoPushDefaultBranch,
        RepoMergeMain,
        RepoMutateCrossRepo,
        SelfModificationActivate,
        SecretAccess,
        PolicyChange,
    };

    public static bool IsProtected(string action) => All.Contains(action);
}

// ── AuthorityRequestContent ───────────────────────────────────────────

/// <summary>Typed content for the authority.requested kernel event.</summary>
public sealed record AuthorityRequestContent(
    string Action,
    ActorId Actor,
    AuthorityLevel Level,
    string Justification,
    IReadOnlyList<EventId> Causes);

public static class AuthorityRequest
{
    /// <summary>
    /// Create record-only authority.requested content for a DF-SOP-0001 protected action.
    /// This does not execute or authorize the side effect.
    /// </summary>
    public static AuthorityRequestContent ProtectedSideEffect(
        string action,
        ActorId actor,
        string justification,
        IReadOnlyList<EventId> causes)
    {
        if (!ProtectedAction.IsProtected(action))
        {
            throw new ArgumentException($"unknown protected action {action}", nameof(action));
        }

        return new AuthorityRequestContent(
            action,
            actor,
            AuthorityLevel.Required,
            justification,
            causes);
    }
}

// ── AuthorityLink ──────────────────────────────────────────────────────

/// <summary>A single link in an authority chain — actor, level, and weight.</summary>
public sealed record AuthorityLink(ActorId Actor, AuthorityLevel Level, Score Weight);

// ── AuthorityResult ────────────────────────────────────────────────────

/// <summary>Result of evaluating authority for an action.</summary>
public sealed record AuthorityResult(
    AuthorityLevel Level,
    Score Weight,
    IReadOnlyList<AuthorityLink> Chain);

// ── AuthorityPolicy ────────────────────────────────────────────────────

/// <summary>Defines the authority requirements for an action pattern.</summary>
public sealed record AuthorityPolicy(
    string Action,
    AuthorityLevel Level,
    Score? MinTrust = null,
    DomainScope? Scope = null);

// ── IAuthorityChain ────────────────────────────────────────────────────

/// <summary>Evaluates authority. Returns weighted authority, not binary permission.</summary>
public interface IAuthorityChain
{
    /// <summary>Evaluate the authority level for an actor performing an action.</summary>
    AuthorityResult Evaluate(ActorId actor, string action);

    /// <summary>Return the authority chain for an actor performing an action.</summary>
    IReadOnlyList<AuthorityLink> Chain(ActorId actor, string action);

    /// <summary>Grant authority from one actor to another in a domain scope.</summary>
    void Grant(ActorId from, ActorId to, DomainScope scope, Score weight);

    /// <summary>Revoke authority from one actor to another in a domain scope.</summary>
    void Revoke(ActorId from, ActorId to, DomainScope scope);
}

// ── DefaultAuthorityChain ──────────────────────────────────────────────

/// <summary>
/// Flat authority model — no delegation chain.
/// All actions default to Notification unless a policy says otherwise.
/// Thread-safe with lock.
/// </summary>
public sealed class DefaultAuthorityChain : IAuthorityChain
{
    private readonly object _lock = new();
    private readonly List<AuthorityPolicy> _policies = new();
    private readonly ITrustModel _trustModel;

    public DefaultAuthorityChain(ITrustModel trustModel)
    {
        _trustModel = trustModel;
    }

    /// <summary>Register an authority policy. Policies are checked in order; first match wins.</summary>
    public void AddPolicy(AuthorityPolicy policy)
    {
        lock (_lock)
        {
            _policies.Add(policy);
        }
    }

    public AuthorityResult Evaluate(ActorId actor, string action)
    {
        AuthorityPolicy policy;
        lock (_lock)
        {
            policy = FindPolicy(action);
        }

        var level = policy.Level;

        // If actor has high enough trust, downgrade Required -> Recommended
        if (level == AuthorityLevel.Required && policy.MinTrust is not null)
        {
            // Build a minimal Actor to query trust — ITrustModel.Score takes Actor
            var metrics = _trustModel.Score(BuildMinimalActor(actor));
            if (metrics.Overall.Value >= policy.MinTrust.Value.Value)
            {
                level = AuthorityLevel.Recommended;
            }
        }

        var link = new AuthorityLink(actor, level, new Score(1.0));

        return new AuthorityResult(level, new Score(1.0), new[] { link });
    }

    public IReadOnlyList<AuthorityLink> Chain(ActorId actor, string action)
    {
        AuthorityPolicy policy;
        lock (_lock)
        {
            policy = FindPolicy(action);
        }

        return new[] { new AuthorityLink(actor, policy.Level, new Score(1.0)) };
    }

    /// <summary>Grant is a no-op in the flat model — would emit an edge.created event in full impl.</summary>
    public void Grant(ActorId from, ActorId to, DomainScope scope, Score weight)
    {
        // No-op stub — flat model has no delegation
    }

    /// <summary>Revoke is a no-op in the flat model — would emit a supersede event in full impl.</summary>
    public void Revoke(ActorId from, ActorId to, DomainScope scope)
    {
        // No-op stub — flat model has no delegation
    }

    // ── Private helpers ────────────────────────────────────────────────

    /// <summary>Find the first matching policy. Must be called under lock.</summary>
    private AuthorityPolicy FindPolicy(string action)
    {
        foreach (var p in _policies)
        {
            if (MatchesAction(p.Action, action))
                return p;
        }

        // Default: Notification level
        return new AuthorityPolicy("*", AuthorityLevel.Notification);
    }

    /// <summary>Build a minimal Actor for trust model queries.</summary>
    private static Actor BuildMinimalActor(ActorId id)
    {
        return new Actor(
            id,
            new PublicKey(new byte[32]),
            "authority-query",
            ActorType.System,
            null,
            0,
            ActorStatus.Active);
    }

    /// <summary>Match an action against a policy pattern. Supports exact, prefix wildcard, and global wildcard.</summary>
    public static bool MatchesAction(string pattern, string action)
    {
        if (pattern == "*")
            return true;

        if (pattern.Length > 0 && pattern[^1] == '*')
        {
            var prefix = pattern[..^1];
            return action.Length >= prefix.Length && action[..prefix.Length] == prefix;
        }

        return pattern == action;
    }
}
